actions/inventory.go

```diff
@@ -353,178 +464,218 @@ func NICs(ctx context.Context, nics []*common.NIC, c NICCollector) error {
-	defer func() {
-		if r := recover(); r != nil {
-			log.Println("recovered from panic in TPM(): ", r)
+// CollectTPMs executes the TPM collector and updates device TPM information
+func (a *InventoryCollectorAction) CollectTPMs(ctx context.Context) error {
+	// nolint:errcheck // deferred method catches a panic, error check not required.
+	defer func() error {
+		if r := recover(); r != nil && a.failOnError {
+			return errors.Wrap(ErrPanic, string(debug.Stack()))
 		}
+
+		return nil
 	}()
 
-	if c == nil {
+	if a.collectors.TPMCollector == nil {
 		return nil
 	}
 
-	ntpms, err := c.TPMs(ctx)
-	if err != nil {
-		return err
+	// skip collector if its been disabled
+	collectorKind, _, _ := a.collectors.TPMCollector.Attributes()
+	if slices.Contains(a.disableCollectorUtilities, collectorKind) {
+		return nil
 	}
 
-	if ntpms == nil {
-		return nil
+	found, err := a.collectors.TPMs(ctx)
+	if err != nil {
+		return err
 	}
 
-	// no tpms identified
-	if len(ntpms) == 0 {
+	if found == nil {
 		return nil
 	}
 
-	// no existing tpms were passed in
-	if len(*tpms) > 0 {
-		changelog, err := diff.Diff(tpms, ntpms)
+	if len(a.device.TPMs) > 0 {
+		changelog, err := diff.Diff(a.device.TPMs, found)
 		if err != nil {
 			return err
 		}
 
-		changelog = vetChanges(changelog)
-		diff.Patch(changelog, tpms)
+		changelog = a.vetChanges(changelog)
+		diff.Patch(changelog, a.device.TPMs)
 	} else {
-		*tpms = append(*tpms, ntpms...)
+		a.device.TPMs = append(a.device.TPMs, found...)
 	}
 
 	return nil
 }
 
-// StorageControllers executes the StorageControllers collector and updates device storage controller data
-func StorageController(ctx context.Context, controllers []*common.StorageController, c StorageControllerCollector) error {
-	defer func() {
-		if r := recover(); r != nil {
-			log.Println("recovered from panic in StorageController(): ", r)
+// CollectStorageControllers executes the StorageControllers collectors and updates device storage controller data
+//
+// nolint:gocyclo // this is fine for now
+func (a *InventoryCollectorAction) CollectStorageControllers(ctx context.Context) error {
+	// nolint:errcheck // deferred method catches a panic, error check not required.
+	defer func() error {
+		if r := recover(); r != nil && a.failOnError {
+			return errors.Wrap(ErrPanic, string(debug.Stack()))
 		}
+
+		return nil
 	}()
 
-	if c == nil {
+	if a.collectors.StorageControllerCollector == nil {
+		return nil
+	}
+
+	// skip collector if its been disabled
+	collectorKind, _, _ := a.collectors.StorageControllerCollector.Attributes()
+	if slices.Contains(a.disableCollectorUtilities, collectorKind) {
 		return nil
 	}
 
-	ncontrollers, err := c.StorageControllers(ctx)
+	found, err := a.collectors.StorageControllers(ctx)
 	if err != nil {
 		return err
 	}
 
-	if len(ncontrollers) == 0 {
+	if len(found) == 0 {
 		return nil
 	}
 
 	// TODO: handle case where the object may not already be present in device.Controllers and needs to be added
```

Comment:
Is this still the case?  If so, do we want to pull in the code added to the Drive collector bit into here?  Maybe into a function that they all can call?

Reply:
This was supposed to be on the TODO comment, not that if... not sure how I fouled that up.

Reply:
@splaspood yep that did cross my mind, I can PR that change after this gets merged though, since the scope of the changes here is quite large already.